### PR TITLE
added monitorTX class & add to orchs list

### PR DESCRIPTION
**Pull request by noaOrMlnx**
**What I did**: added class in orchagent named MonitorTXOrch and added it to the orchs list in orchdaemon class

**Why I did it**

**How I verified it**

**Details if related**

### CODE REVIEW

File: orchagent/monitorTXorch.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+
+enum PortState {OK, NOT_OK, UNKNOWN};
+
+
+class MonitorTXOrch : public Orch
+{
+    
+    public:
+        //MonitorTXOrch(DBConnector* ConfigDB, DBConnector* StateDB);
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.h]: remove commented code

File: orchagent/monitorTXorch.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+// default pooling period & threshold
+#define TX_DEFAULT_POOLING_PERIOD 10
+#define TX_DEFAULT_THRESHOLD 100
+
+
+extern "C" {
+    #include "sai.h"
+}
+
+
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.h]: extra empty line

File: orchagent/monitorTXorch.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+enum PortState {OK, NOT_OK, UNKNOWN};
+
+
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.h]: extra empty line

File: orchagent/monitorTXorch.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        // counters table - taken from redis 
+        shared_ptr<swss::DBConnector> m_countersDB = nullptr;
+        shared_ptr<swss::Table> m_countersTable = nullptr;
+        shared_ptr<Table> m_countersPortNameMap = nullptr;
+
+        // state table
+        Table m_stateTxErrTable;
+
+        // config table
+        // Table m_cfgTxErrTable;
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.h]: remove commented code

File: orchagent/monitorTXorch.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+class MonitorTXOrch : public Orch
+{
+    
+    public:
+        //MonitorTXOrch(DBConnector* ConfigDB, DBConnector* StateDB);
+        MonitorTXOrch(TableConnector configDBTConnector, TableConnector stateDBTConnector);
+        virtual void doTask(SelectableTimer &timer);
+        virtual void doTask(Consumer &consumer);
+
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.h:41]: MonitorTXOrch inherits from virtual class, it is more correct to have virtual destructor for MonitorTXOrch

File: orchagent/monitorTXorch.h [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        map<sai_object_id_t, uint64_t> m_TX_ERR_stat_by_port;
+        map<sai_object_id_t, uint64_t> m_prevTXCounters;
+        map<sai_object_id_t, uint64_t> m_currTXCounters;
+
+        map<sai_object_id_t, PortState> m_portsState;
+
+        uint64_t m_threshold = TX_DEFAULT_THRESHOLD;
+        int m_poolingPeriod = TX_DEFAULT_POOLING_PERIOD;
+
+        bool isPortMapInitialized;
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.h]: You use a lot of default intialized values for member fields, which is good.
Use the same for isPortMapInitialized and initialize by default to false

File: orchagent/monitorTXorch.h [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+        bool createPortNameMap();
+        void initTimer();
+        bool getTXStatistics();
+        bool getTXCountersById(sai_object_id_t portID);
+        void updatePortsState();
+        // void insertStateToDB(sai_object_id_t portID, PortState state);
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.h]: remove commented code

File: orchagent/monitorTXorch.h [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        bool handleSetCommand(const string& key, const vector<FieldValueTuple>& data);
+
+
+};
+
+#endif
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.h]: common practice is to insert newline at the end of file

File: orchagent/monitorTXorch.h [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        // state table
+        Table m_stateTxErrTable;
+
+        // config table
+        // Table m_cfgTxErrTable;
+
+        map<sai_object_id_t, string> m_portsStringsMap;
+
+        // counetrs of tx errors
+        map<sai_object_id_t, uint64_t> m_TX_ERR_stat_by_port;
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.h]: inconsistent naming, suggest to use m_txErrStatByPort

File: orchagent/monitorTXorch.h [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        map<sai_object_id_t, uint64_t> m_prevTXCounters;
+        map<sai_object_id_t, uint64_t> m_currTXCounters;
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.h]: How this m_currTXCounters is different from m_TX_ERR_stat_by_port?

File: orchagent/monitorTXorch.cpp [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+extern PortsOrch* gPortsOrch;
+// extern sai_port_api_t * sai_port_api;
+
+static const array<string, STATES_NUMBER> stateNames = {"OK", "NOT_OK", "UNKNOWN"};
+static const string counterName = "SAI_PORT_STAT_IF_OUT_ERRORS";
+static const string portNameMap = "COUNTERS_PORT_NAME_MAP";
+
+
+// constructor
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: basically useless comment

File: orchagent/monitorTXorch.cpp [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+MonitorTXOrch::MonitorTXOrch(TableConnector configDBTConnector, TableConnector stateDBTConnector):
+    // listens to configdb changes
+    Orch(configDBTConnector.first, configDBTConnector.second),
+
+    m_countersDB(new DBConnector(COUNTERS_DB, DBConnector::DEFAULT_UNIXSOCKET, 0)),
+    m_countersTable(new Table(m_countersDB.get(), "COUNTERS")),
+    m_countersPortNameMap(new Table(m_countersDB.get(), COUNTERS_PORT_NAME_MAP)),
+    // connector to state db
+    m_stateTxErrTable(stateDBTConnector.first, stateDBTConnector.second)    
+    // m_cfgTxErrTable(configDBTConnector.first, configDBTConnector.second)
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: remove commented code

File: orchagent/monitorTXorch.cpp [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+
+{
+    SWSS_LOG_ENTER();
+    SWSS_LOG_NOTICE("creating monitor tx orch ...");
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: too verbose logging, it should be SWSS_LOG_DEBUG

File: orchagent/monitorTXorch.cpp [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+    initTimer();
+    SWSS_LOG_NOTICE("after init timer");
+    isPortMapInitialized = false;
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: Not required if you initialize in class definition

File: orchagent/monitorTXorch.cpp [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+    m_countersPortNameMap(new Table(m_countersDB.get(), COUNTERS_PORT_NAME_MAP)),
+    // connector to state db
+    m_stateTxErrTable(stateDBTConnector.first, stateDBTConnector.second)    
+    // m_cfgTxErrTable(configDBTConnector.first, configDBTConnector.second)
+
+{
+    SWSS_LOG_ENTER();
+    SWSS_LOG_NOTICE("creating monitor tx orch ...");
+    initTimer();
+    SWSS_LOG_NOTICE("after init timer");
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: Not required log

File: orchagent/monitorTXorch.cpp [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+    isPortMapInitialized = false;
+ 
+}
+
+bool MonitorTXOrch::createPortNameMap() {
+    SWSS_LOG_ENTER();
+    SWSS_LOG_NOTICE("all ports are ready");
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: General comment - NOTICE is too verbose for such log message, either delete or use SWSS_LOG_DEBUG

File: orchagent/monitorTXorch.cpp [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+    map<string, Port> &portsList =  gPortsOrch->getAllPorts();
+    for (auto &entry : portsList) {
+        string name = entry.first;
+        SWSS_LOG_NOTICE("PORT NAME IS %s", name.c_str());
+        Port p = entry.second;
+        if (p.m_type == Port::PHY) {
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: convert logic to more readable flow:

```c++
if (p.m_type != Port::PHY)
{
   continue;
}
// procceed.. 
```

File: orchagent/monitorTXorch.cpp [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+    }
+    SWSS_LOG_NOTICE("exiting the createportname func");
+    return true;
+}
+
+
+/*
+initialize timer with default pooling period
+*/
+void MonitorTXOrch::initTimer() {
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: General comment about sonic code style - ```{``` should be on new line

File: orchagent/monitorTXorch.cpp [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+    if (key == POOLING_PERIOD_KEY) {
+        for (auto element : data) {
+            const auto &field = fvField(element);
+            const auto &value = fvValue(element);
+
+            // change the interval and reset timer
+
+            if (field == "value") {
+                SWSS_LOG_NOTICE("changing pooling period to %s", value.c_str());
+                // m_poolingPeriod = chrono::seconds(to_uint<uint32_t>(value));
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: remove commented code

File: orchagent/monitorTXorch.cpp [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+                m_poolingPeriod = stoi(value);
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp:109]: value is user input.
Setting the value to some garbage will crash orchagent.
Here you need 
```c++
try
{
}
catch (...)
{
}
```

File: orchagent/monitorTXorch.cpp [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+                auto interv = timespec { .tv_sec = m_poolingPeriod, .tv_nsec = 0 };
+                m_timer->setInterval(interv);
+                m_timer->reset();
+            }
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp:118]: Add ``` else ``` block with log message about unknown ```field``` value

File: orchagent/monitorTXorch.cpp [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+    } else if (key == THRESHOLD_KEY) {
+
+        for (auto element : data) {
+            const auto &field = fvField(element);
+            const auto &value = fvValue(element);
+
+            if (field == "value") {
+                SWSS_LOG_NOTICE("changing threshold to %s", value.c_str());
+                m_threshold = stoi(value);
+            }
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp:145]: same else block is needed here

File: orchagent/monitorTXorch.cpp [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+
+    // change threshold val
+    } else if (key == THRESHOLD_KEY) {
+
+        for (auto element : data) {
+            const auto &field = fvField(element);
+            const auto &value = fvValue(element);
+
+            if (field == "value") {
+                SWSS_LOG_NOTICE("changing threshold to %s", value.c_str());
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: This kind of log messages are good to be with NOTICE level

File: orchagent/monitorTXorch.cpp [hunk 24]
```diff
@@ -0,0 +1,300 @@
...
+            if (!isSet) {
+                SWSS_LOG_ERROR("set has not succeed");
+            }
+        } 
+        else 
+        {
+            SWSS_LOG_ERROR("Unknown operation type %s", op.c_str());
+        }
+
+        consumer.m_toSync.erase(it++);
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: This line of code worries me a bit.
An order is that first the current entry pointed by ```it``` will be erased and the the ```it``` will be incremented to the next one.
I believe this is undefined behavior when you operate on iterator which points to removed entry from map
A correct and simpler way is to:
```c++
it = consumer.m_toSync.erase(it)
```

File: orchagent/monitorTXorch.cpp [hunk 25]
```diff
@@ -0,0 +1,300 @@
...
+
+    }
+}
+
+// void MonitorTXOrch::doTask(Consumer &consumer) {
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: remove commented code

File: orchagent/monitorTXorch.cpp [hunk 26]
```diff
@@ -0,0 +1,300 @@
...
+        m_currTXCounters.emplace(currPortId, (totalTxErr - prev));
+
+        if (m_currTXCounters.find(currPortId)->second > m_threshold){
+
+            currState = NOT_OK;
+        }
+        // SWSS_LOG_NOTICE("port %s state is %s", sai_serialize_object_id(currPortId), stateNames[currState]);
+        SWSS_LOG_NOTICE("port s state is s");
+
+        m_portsState.emplace(currPortId, currState);
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: In state DB there should be no ID, it should be alias

File: orchagent/monitorTXorch.cpp [hunk 27]
```diff
@@ -0,0 +1,300 @@
...
+//     if (!m_stateTxErrTable.set(sai_serialize_object_id(currPortId), fieldValuesVector)) {
+//         SWSS_LOG_ERROR("set to state db has not succeed");
+//     }
+// }
+
+void MonitorTXOrch::updatePortsState() {
+    SWSS_LOG_ENTER();
+    SWSS_LOG_NOTICE("inside update port state");
+    
+    // TODO: check if need to flush here  - clear table before inserting new states.
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: your new state will override existing so no need to clear or flush

File: orchagent/monitorTXorch.cpp [hunk 28]
```diff
@@ -0,0 +1,300 @@
...
+        SWSS_LOG_NOTICE("port s state is s");
+
+        m_portsState.emplace(currPortId, currState);
+
+        vector<FieldValueTuple> fieldValuesVector;
+        fieldValuesVector.emplace_back(STATE_DB_TX_STATE, stateNames[currState]);
+  
+        m_stateTxErrTable.set(oidStr, fieldValuesVector);
+        SWSS_LOG_NOTICE("Set status s to port s");
+
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: extra newline

File: orchagent/monitorTXorch.cpp [hunk 29]
```diff
@@ -0,0 +1,300 @@
...
+        m_stateTxErrTable.set(oidStr, fieldValuesVector);
+        SWSS_LOG_ERROR("Cannot take information from table for port %s", oidStr.c_str());
+        return false;
+    }
+    m_TX_ERR_stat_by_port.emplace(portID, stoul(strValue));
+    SWSS_LOG_NOTICE("emplaced port %s in map of stat", oidStr.c_str());
+    return true; 
+}
+
+// void MonitorTXOrch::insertStateToDB(sai_object_id_t portID, PortState state) {
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: remove commented code

File: orchagent/monitorTXorch.cpp [hunk 30]
```diff
@@ -0,0 +1,300 @@
...
+
+    SWSS_LOG_NOTICE("getting tx-err counters to specific port %s", oidStr.c_str());
+
+    if (!m_countersTable->hget(oidStr, counterName, strValue)) {
+        SWSS_LOG_NOTICE("Error reading counters table");
+        // insert unknown state to DB
+
+        vector<FieldValueTuple> fieldValuesVector;
+        fieldValuesVector.emplace_back(STATE_DB_TX_STATE, "UNKNOWN");
+        m_stateTxErrTable.set(oidStr, fieldValuesVector);
```
Comment (stepanblyschak) [inline orchagent/monitorTXorch.cpp]: there are two places where you set the state, can you have a small routine that sets the state so this part of code will be reused.

**Review by stepanblyschak — CHANGES_REQUESTED** [review]

File: orchagent/monitorTXorch.h (hunk 10 above)
Comment (noaOrMlnx) [reply orchagent/monitorTXorch.h]: m_txErrStatByPort is the data taken from counters - all tx errors until now. 
the current map contains the subtraction between the total count and the previous count. (so the check of threshold will be in the period defined).

**Pull request closed without merging**